### PR TITLE
JavaScript - Choices Recognizer

### PR DESCRIPTION
**Pull request by pcostantini**
- Refactor base and common interfaces by moving them to recognizers-text-base package
- New Choices Recognizer and Boolean Recognizer (recognizer-text-choices package)
  - Boolean Recognizer logic extracted from BotBuilder's EntityRecognizer
  - Regex extracted to new [Choices definition file](https://github.com/southworkscom/Recognizers-Text/blob/066a913189437151dd62bc6aa87260783d5d336c/Patterns/English/English-Choices.yaml)
- New Specs and Test Runner for Boolean Recognizer

### CODE REVIEW

File: Patterns/English/English-Choices.yaml [hunk 1]
```diff
@@ -0,0 +1,9 @@
+﻿---
+LangMarker: Eng
+#Boolean Recognizers
+TokenizerRegex: !simpleRegex
+  def: '[^\w\d]'
+TrueRegex: !simpleRegex
+  def: \b(y(e[sp])?|sure|ok|\uD83D\uDC4D|\uD83D\uDC4C)\b
```
Comment (Stevenic) [inline Patterns/English/English-Choices.yaml]: add "true" to this

File: Patterns/English/English-Choices.yaml [hunk 2]
```diff
@@ -0,0 +1,9 @@
...
+FalseRegex: !simpleRegex
+  def: \b(n(o(pe)?)?|\uD83D\uDC4E|\u270B|\uD83D\uDD90)\b
```
Comment (Stevenic) [inline Patterns/English/English-Choices.yaml]: add "false" to this

File: JavaScript/packages/recognizers-choices/src/resources/englishChoices.ts [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+//     This code was generated by a tool.
+//     Changes to this file may cause incorrect behavior and will be lost if
+//     the code is regenerated.
+// </auto-generated>
+// ------------------------------------------------------------------------------
+
+export namespace EnglishChoices {
+	export const LangMarker = 'Eng';
+	export const TokenizerRegex = `[^\\w\\d]`;
+	export const TrueRegex = `\\b(y(e[sp])?|sure|ok|\\uD83D\\uDC4D|\\uD83D\\uDC4C)\\b`;
```
Comment (Stevenic) [inline JavaScript/packages/recognizers-choices/src/resources/englishChoices.ts]: >\\b(y(e[sp])?|sure|ok|\\uD83D\\uDC4D|\\uD83D\\uDC4C)\\b` [](start = 27, length = 56)

this should include "true"

File: JavaScript/packages/recognizers-choices/src/resources/englishChoices.ts [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+	export const FalseRegex = `\\b(n(o(pe)?)?|\\uD83D\\uDC4E|\\u270B|\\uD83D\\uDD90)\\b`;
```
Comment (Stevenic) [inline JavaScript/packages/recognizers-choices/src/resources/englishChoices.ts]: >\\b(n(o(pe)?)?|\\uD83D\\uDC4E|\\u270B|\\uD83D\\uDD90)\\b [](start = 28, length = 56)

this should include "false"

**Comment by johnataylor** [comment]
For now could we just take the refactoring and the Boolean recognizer from this pull request.

Because Steve is working on a full choice recognizer and will do that work in the BotBuilder repo at least for now.

However, we definitely want the Boolean recognizer and with Steve's review feedback.

So if we could also rename to reflect that narrower scope that would be great.

File: Patterns/English/English-Choices.yaml (hunk 1 above)
Comment (johnataylor) [reply Patterns/English/English-Choices.yaml]: The regex would be easier to read and maintain if it was simpler. Specifically avoiding the grouping and flattening the overall structure. So for the true case y|yes|yep reads better than y(e[sp])? same feedback applies for the false case.

Comment (enzocanoo) [reply Patterns/English/English-Choices.yaml]: Changed to improve readability

**Comment by enzocanoo** [comment]
- Changed the meta-module name to `recognizers-text-suite`
- Changed the base-module name to `recognizers-text`
- Changed the new choices module name to `recognizers-text-options`

**Comment by pcostantini** [comment]
Hi @tellarin,
@enzocanoo applied the refactoring feedback (modules renaming) and also John's feedback about having just the Boolean Recognizer (no base ChoiceRecognizer).
Once Steve completes his ChoiceRecognizer, we can re-visit refactoring any interfaces and integrating new code. As-is, this PR is now complete.
Regards,
-PabloC.

**Pull request merged**